sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -883,6 +893,274 @@ def _export_files(self, bq):
     return table.schema, metadata_list
 
 
+class _CustomBigQueryStorageSourceBase(BoundedSource):
+  """A base class for BoundedSource implementations which read from BigQuery
+  using the BigQuery Storage API.
+
+  Args:
+    table (str, TableReference): The ID of the table. The ID must contain only
+      letters ``a-z``, ``A-Z``, numbers ``0-9``, or underscores ``_``  If
+      **dataset** argument is :data:`None` then the table argument must
+      contain the entire table reference specified as:
+      ``'PROJECT:DATASET.TABLE'`` or must specify a TableReference.
+    dataset (str): Optional ID of the dataset containing this table or
+      :data:`None` if the table argument specifies a TableReference.
+    project (str): Optional ID of the project containing this table or
+      :data:`None` if the table argument specifies a TableReference.
+    selected_fields (List[str]): Optional List of names of the fields in the
+      table that should be read. If empty, all fields will be read. If the
+      specified field is a nested field, all the sub-fields in the field will be
+      selected. The output field order is unrelated to the order of fields in
+      selected_fields.
+    row_restriction (str): Optional SQL text filtering statement, similar to a
+      WHERE clause in a query. Aggregates are not supported. Restricted to a
+      maximum length for 1 MB.
+  """
+
+  # The maximum number of streams which will be requested when creating a read
+  # session, regardless of the desired bundle size.
+  MAX_SPLIT_COUNT = 10000
+  # The minimum number of streams which will be requested when creating a read
+  # session, regardless of the desired bundle size. Note that the server may
+  # still choose to return fewer than ten streams based on the layout of the
+  # table.
+  MIN_SPLIT_COUNT = 10
+
+  def __init__(
+      self,
+      table: Union[str, TableReference],
+      dataset: Optional[str] = None,
+      project: Optional[str] = None,
+      selected_fields: Optional[List[str]] = None,
+      row_restriction: Optional[str] = None,
+      use_fastavro: Optional[bool] = None,
```

Comment:
What would be circumstances when the user need to use python avro reader? The python avro library is so slow that I am not sure what would be use case for it.  

> The current Python avro package is dog slow.

> On a test case of about 10K records, it takes about 14sec to iterate over all of them. In comparison the JAVA avro SDK does it in about 1.9sec.

> fastavro is an alternative implementation that is much faster. It iterates over the same 10K records in 2.9sec, and if you use it with PyPy it’ll do it in 1.5sec (to be fair, the JAVA benchmark is doing some extra JSON encoding/decoding).

https://fastavro.readthedocs.io/en/latest/

Reply:
Updated `estimate_size()` which should improve autoscaling in [theory]. There should be further improvements in autoscaling once we add support dynamic work rebalancing. Additionally, I intend to remove the non-fastavro parsing in a follow-up PR where the Storage API source and the core `avroio` source are combined together.

[theory]:  https://cloud.google.com/dataflow/docs/guides/deploying-a-pipeline#batch-autoscaling

Reply:
I can confirm that this actually worked. An autoscaller scaled all workers from the start.

Reply:
That's great! Is performance now comparable with that of an EXPORT job?

Reply:
Performance even better than Export job.  Caveat: I did only a small number testing, but overall performance right now on par or even better compare to export job.